src/OrchardCore.Modules/OrchardCore.Roles/Startup.cs

```diff
@@ -35,12 +35,13 @@ public Startup(IOptions<AdminOptions> adminOptions)
 
         public override void ConfigureServices(IServiceCollection services)
         {
+            services.AddScoped<RoleStore>();
+            services.TryAddScoped<IRoleClaimStore<IRole>>(sp => sp.GetRequiredService<RoleStore>());
+            services.AddScoped<IRoleStore<IRole>>(sp => sp.GetRequiredService<RoleStore>());
+
             services.TryAddScoped<RoleManager<IRole>>();
```

Comment:
If I follow the pattern here we are using some `AddScoped` and in `AddUsers()` we are using `TryAddScoped()`, but this implies that the Users startup should run after the Roles startup, see below.

Line 38 no problem as we are registering different implementations.

Line 39 I think you meant `AddScoped()` in place of `TryAddScoped()`, but I think that for line 39 and 40 this is not the right  pattern, see below.

Line 42 okay as we are trying to register in both places the same `RoleManager` implementation.

For line 38 and 39, because Users and Roles have no dependency on each other, **I think it works** because the Users startup runs last **but only because Users is after Roles in an alphabetic order**.

When we can use dependencies a dependent can just re-register a service because its startup runs after its dependency, more complex when services are resolved by `IEnumerable<T>` but there are helper extensions.

When there is no manifest dependencies we can tweak the Order property of the startup, eventually a new startup class dedicated to a set of services if we don't want to change the order of other services.

 In our case I think better to still use `TryAddScoped()` in `AddUsers()` to register the default implementations, and here use `Replace()` (maybe `ReplaceAll()` will see) to replace the defaults (less registrations) if already registered.

Another solution would be to register both `RoleStore` and `NullRoleStore` with the same new interface e.g. `IRoleAndClaimStore`, then the implementation factory delegates would resolve this new interface, so here we would only need to replace the registration of this new interface, no need to replace `IRoleClaimStore` and `IRoleStore`.

Reply:
@jtkech I am not aware of a `ReplaceAll()` method. But I can do this for a safer option

```
services.RemoveAll<IRoleClaimStore<IRole>>();
services.RemoveAll<IRoleStore<IRole>>();
services.AddScoped<IRoleClaimStore<IRole>>(sp => sp.GetRequiredService<RoleStore>());
services.AddScoped<IRoleStore<IRole>>(sp => sp.GetRequiredService<RoleStore>());
```
This way we always ensure that only one instance is registered when when roles is called.

Reply:
@jtkech I agree with you line 39 should have been `AddScoped<>` instead of `TryAddScoped<>`. I also agree that `Replace()` may be a safer option here. I made the change using Replace, If you see no additional issues, please merge it so you can complete test out #13040 with this new change.

```
services.TryAddScoped<IRoleClaimStore<IRole>>(sp => sp.GetRequiredService<RoleStore>());
            services.AddScoped<IRoleStore<IRole>>(sp => sp.GetRequiredService<RoleStore>());
```

Reply:
Okay, I'm just thinking about the new e.g. `IRoleAndClaimStore` interface suggestion that could be used to register both `NullRoleStore` and `RoleStore` and then used in the fatory delegates, so that we only need to replace this registration in the `OC.Roles` startup.

For example in `AddUsers()` somethig like this.

    s.TryAddScoped<IRoleAndClaimStore, NullRoleStore>(); // TryAddScoped() used here.
    s.AddScoped<IRoleClaimStore<IRole>>(sp => sp.GetRequiredService<IRoleAndClaimStore>());
    s.AddScoped<IRoleStore<IRole>>(sp => sp.GetRequiredService<IRoleAndClaimStore>());

Then in the `OC.Roles` startup only.

    s.Replace(ServiceDescriptor.Scoped<IRoleAndClaimStore, RoleStore>());

Reply:
Anyway, that"s okay like this, I will merge it ;)